### PR TITLE
Move to middleware v2

### PR DESCRIPTION
**Pull request by benjih**
Middleware v2 is a big update to how we handle middleware. Instead of pointing Hoverfly to a script, we now provide Hoverfly with the script. Hoverfly will now manage the script itself.

The benefit of doing this is to
1. Avoid any issues with paths when starting Hoverfly in one directory and then changing directory to set the middleware
2. Allow middleware to be set remotely

This update also tidies up the middleware part of Hoverfly as well as the web interface for setting both local binary middlewares and middlewares over HTTP.

### CODE REVIEW

File: core/handlers/v2/hoverfly_middleware_handler.go [hunk 1]
```diff
@@ -2,16 +2,19 @@ package v2
...
+	"net/http"
+
 	"github.com/SpectoLabs/hoverfly/core/handlers"
 	"github.com/codegangsta/negroni"
 	"github.com/go-zoo/bone"
-	"io/ioutil"
-	"net/http"
 )
 
 type HoverflyMiddleware interface {
```
Comment (benjih) [inline core/handlers/v2/hoverfly_middleware_handler.go]: This interface needs tidying up and removing the v1 functions.

File: core/handlers/v2/hoverfly_middleware_handler_test.go [hunk 2]
```diff
@@ -4,20 +4,28 @@ import (
...
 	"net/http"
 	"testing"
+
+	. "github.com/onsi/gomega"
 )
 
 type HoverflyMiddlewareStub struct {
+	Binary     string
+	Script     string
 	Middleware string
```
Comment (benjih) [inline core/handlers/v2/hoverfly_middleware_handler_test.go]: This old field needs removing.

File: core/hoverfly_service.go [hunk 3]
```diff
@@ -102,13 +102,15 @@ func (hf Hoverfly) SetMiddleware(middleware string) error {
...
 	newMiddleware := Middleware{}
 
-	if binary == "" && script == "" {
+	if binary == "" && script == "" && remote == "" {
 		hf.Cfg.Middleware = newMiddleware
 		return nil
-	} else if binary == "" {
+	}
+
+	if binary == "" && script != "" {
```
Comment (danielbryantuk) [inline core/hoverfly_service.go]: Wat?? I think the binary/script check be reversed ;-)

File: core/handlers/v2/views.go [hunk 4]
```diff
@@ -17,7 +17,10 @@ type UsageView struct {
 }
 
 type MiddlewareView struct {
+	Binary     string `json:"binary"`
 	Middleware string `json:"middleware"`
```
Comment (benjih) [inline core/handlers/v2/views.go]: This old field needs removing.

File: core/middleware.go [hunk 5]
```diff
@@ -88,7 +105,7 @@ func (this *Middleware) Execute(pair models.RequestResponsePair) (models.Request
 		this.FullCommand = this.Binary + " " + this.Script.Name()
 	}
 
-	if this.isLocal() {
+	if this.isLocal() && this.Remote == "" {
```
Comment (danielbryantuk) [inline core/middleware.go]: This looks like a 'double-barrel' check?

File: core/middleware_test.go [hunk 6]
```diff
@@ -398,3 +432,92 @@ func Test_Middleware_Execute_RunsMiddlewareCorrectly(t *testing.T) {
...
+
+	unit := Middleware{}
+
+	err := unit.SetRemote("")
+	Expect(err).To(BeNil())
+
+	Expect(unit.Remote).To(Equal(""))
+}
+
+// This test is relevant, but is broken for now, it shall be uncommented soon
```
Comment (danielbryantuk) [inline core/middleware_test.go]: hmmm....

File: core/models_test.go [hunk 7]
```diff
@@ -88,9 +88,29 @@ func TestRequestBodySentToMiddleware(t *testing.T) {
 	requestDetails, err := models.NewRequestDetailsFromHttpRequest(req)
 	Expect(err).To(BeNil())
 
-	dbClient.Cfg.Middleware.FullCommand = "./examples/middleware/reflect_body/reflect_body.py"
+	err = dbClient.Cfg.Middleware.SetBinary("python")
+	Expect(err).To(BeNil())
+
+	err = dbClient.Cfg.Middleware.SetScript("#!/usr/bin/env python\n" +
```
Comment (danielbryantuk) [inline core/models_test.go]: Move to constant/var?

File: core/synthesize_test.go [hunk 8]
```diff
@@ -54,7 +74,7 @@ func TestSynthesizeMiddlewareFailure(t *testing.T) {
 	Expect(err).To(BeNil())
 
 	middleware := &Middleware{
-		FullCommand: "./examples/middleware/this_is_not_there.py",
+		Script: nil,
```
Comment (benjih) [inline core/synthesize_test.go]: Should this be nil?

**Comment by danielbryantuk** [comment]
LGTM!

File: core/handlers/v2/hoverfly_middleware_handler.go (hunk 1 above)
Comment (benjih) [inline core/handlers/v2/hoverfly_middleware_handler.go]: Cleaned up this interface.

File: core/handlers/v2/hoverfly_middleware_handler_test.go (hunk 2 above)
Comment (benjih) [inline core/handlers/v2/hoverfly_middleware_handler_test.go]: Cleaned up this interface.

File: core/handlers/v2/views.go (hunk 4 above)
Comment (benjih) [inline core/handlers/v2/views.go]: Old field removed.

File: core/middleware.go (hunk 5 above)
Comment (benjih) [inline core/middleware.go]: Double barrel check removed.

File: core/middleware_test.go (hunk 6 above)
Comment (benjih) [inline core/middleware_test.go]: Test uncommented.

File: core/models_test.go (hunk 7 above)
Comment (benjih) [inline core/models_test.go]: Done.

File: core/synthesize_test.go (hunk 8 above)
Comment (benjih) [inline core/synthesize_test.go]: Done

**Pull request merged**